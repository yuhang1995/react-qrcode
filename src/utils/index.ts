export * from './function';
export * from './refs';
