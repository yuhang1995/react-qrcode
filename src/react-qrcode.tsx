import * as React from 'react';
import {
    toDataURL,
    toCanvas,
    QRCodeRenderersOptions,
    QRCodeToDataURLOptions,
} from 'qrcode';

import { mergeRefs } from './utils';

export interface ImageProps {
    text: string;
    options?: QRCodeToDataURLOptions;
}

export const Image = React.forwardRef<HTMLImageElement, ImageProps>(
    (props, propsRef) => {
        const { text, options = {}, ...rest } = props;
        const inputRef = React.useRef(null);

        const ref = mergeRefs(inputRef, propsRef);

        const [src, setSrc] = React.useState('');

        React.useEffect(() => {
            if (inputRef.current) {
                toDataURL(text, options, function(error: unknown, url: string) {
                    if (error) {
                        throw error;
                    }
                    setSrc(url);
                });
            }
        }, [text, options]);

        return <img ref={ref} src={src} {...rest} />;
    }
);

export interface CanvasProps {
    text: string;
    options?: QRCodeRenderersOptions;
}

export const Canvas = React.forwardRef<HTMLCanvasElement, CanvasProps>(
    (props, propsRef) => {
        const { text, options = {}, ...rest } = props;
        const canvasRef = React.useRef(null);

        const ref = mergeRefs(canvasRef, propsRef);

        React.useEffect(() => {
            if (canvasRef.current) {
                toCanvas(canvasRef.current, text, options, function(
                    error: unknown
                ) {
                    if (error) {
                        throw error;
                    }
                });
            }
        }, [text, options]);

        return <canvas ref={ref} {...rest} />;
    }
);
