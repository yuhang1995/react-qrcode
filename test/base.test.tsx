import React from 'react';
// import * as ReactDOM from 'react-dom/client';
import { Image, ImageProps, Canvas } from '../src';
import renderer from 'react-test-renderer';

const BASIC_PROPS = {
    text: 'https://github.com/yuhang1995/react-qrcode.git',
};

const BASIC_IMAGE: ImageProps = {
    ...BASIC_PROPS,
    options: {
        type: 'image/png',
    },
};

const BASIC_CANVAS = {
    ...BASIC_PROPS,
    options: {},
};

const TEST_TEXTS = [
    { text: '1234567890' },
    { text: 'single byte emoji ✅' },
    { text: 'double byte emoji 👌' },
    { text: 'four byte emoji 👌🏽' },
    { text: '你好' },
];

const TEST_LEVEL = ['L', 'H', 'M'];

describe('Image rendering', () => {
    test('renders basic Image correctly', () => {
        const tree = renderer.create(<Image {...BASIC_IMAGE} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test.each(TEST_TEXTS)('renders Image variation (%o) correctly', text => {
        const options = {
            ...BASIC_IMAGE,
            ...text,
        };
        const tree = renderer.create(<Image {...options} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test.each(TEST_LEVEL)('renders Image variation (%o) correctly', level => {
        const options = {
            ...BASIC_IMAGE,
            option: {
                errorCorrectionLevel: level,
            },
        };
        const tree = renderer.create(<Image {...options} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe('Canvas rendering', () => {
    test('renders basic Canvas correctly', () => {
        const tree = renderer.create(<Canvas {...BASIC_CANVAS} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test.each(TEST_TEXTS)('renders Canvas variation (%o) correctly', text => {
        const options = {
            ...BASIC_CANVAS,
            ...text,
        };
        const tree = renderer.create(<Canvas {...options} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test.each(TEST_LEVEL)('renders Canvas variation (%o) correctly', level => {
        const options = {
            ...BASIC_CANVAS,
            option: {
                errorCorrectionLevel: level,
            },
        };
        const tree = renderer.create(<Canvas {...options} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
