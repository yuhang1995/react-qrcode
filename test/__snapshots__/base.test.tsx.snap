// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Canvas rendering renders Canvas variation ('H') correctly 1`] = `
<canvas
  option={
    Object {
      "errorCorrectionLevel": "H",
    }
  }
/>
`;

exports[`Canvas rendering renders Canvas variation ('L') correctly 1`] = `
<canvas
  option={
    Object {
      "errorCorrectionLevel": "L",
    }
  }
/>
`;

exports[`Canvas rendering renders Canvas variation ('M') correctly 1`] = `
<canvas
  option={
    Object {
      "errorCorrectionLevel": "M",
    }
  }
/>
`;

exports[`Canvas rendering renders Canvas variation ({ text: '1234567890' }) correctly 1`] = `<canvas />`;

exports[`Canvas rendering renders Canvas variation ({ text: 'double byte emoji 👌' }) correctly 1`] = `<canvas />`;

exports[`Canvas rendering renders Canvas variation ({ text: 'four byte emoji 👌🏽' }) correctly 1`] = `<canvas />`;

exports[`Canvas rendering renders Canvas variation ({ text: 'single byte emoji ✅' }) correctly 1`] = `<canvas />`;

exports[`Canvas rendering renders Canvas variation ({ text: '你好' }) correctly 1`] = `<canvas />`;

exports[`Canvas rendering renders basic Canvas correctly 1`] = `<canvas />`;

exports[`Image rendering renders Image variation ('H') correctly 1`] = `
<img
  option={
    Object {
      "errorCorrectionLevel": "H",
    }
  }
  src=""
/>
`;

exports[`Image rendering renders Image variation ('L') correctly 1`] = `
<img
  option={
    Object {
      "errorCorrectionLevel": "L",
    }
  }
  src=""
/>
`;

exports[`Image rendering renders Image variation ('M') correctly 1`] = `
<img
  option={
    Object {
      "errorCorrectionLevel": "M",
    }
  }
  src=""
/>
`;

exports[`Image rendering renders Image variation ({ text: '1234567890' }) correctly 1`] = `
<img
  src=""
/>
`;

exports[`Image rendering renders Image variation ({ text: '1234567890' }) correctly 2`] = `
<img
  src=""
/>
`;

exports[`Image rendering renders Image variation ({ text: 'double byte emoji 👌' }) correctly 1`] = `
<img
  src=""
/>
`;

exports[`Image rendering renders Image variation ({ text: 'double byte emoji 👌' }) correctly 2`] = `
<img
  src=""
/>
`;

exports[`Image rendering renders Image variation ({ text: 'four byte emoji 👌🏽' }) correctly 1`] = `
<img
  src=""
/>
`;

exports[`Image rendering renders Image variation ({ text: 'four byte emoji 👌🏽' }) correctly 2`] = `
<img
  src=""
/>
`;

exports[`Image rendering renders Image variation ({ text: 'single byte emoji ✅' }) correctly 1`] = `
<img
  src=""
/>
`;

exports[`Image rendering renders Image variation ({ text: 'single byte emoji ✅' }) correctly 2`] = `
<img
  src=""
/>
`;

exports[`Image rendering renders Image variation ({ text: '你好' }) correctly 1`] = `
<img
  src=""
/>
`;

exports[`Image rendering renders Image variation ({ text: '你好' }) correctly 2`] = `
<img
  src=""
/>
`;

exports[`Image rendering renders basic Image correctly 1`] = `
<img
  src=""
/>
`;
